### ISSUE TITLE
configure appends "-O2" option to `CFLAGS`

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1727]
Here is my output when I run `configure` (Msys2, Windows 10):
```
CFLAGS:               -ffast-math -funroll-loops -fomit-frame-pointer -O3  -g -O2
```
This is bad because `-O2` will override `-O3`, preventing lots of optimizations.

Note that my `CFLAGS` environment variable is empty (verified by running `echo $CFLAGS`) and I don't see this flags in `configure.ac`, so I guess this is somehow added by automake.

We *prepend* our flags to `CFLAGS`, `CPPFLAGS`, etc. because we want to allow the user to override them, so it is really unfortunate if automake decides to fill `CFLAGS` with some default options.

My current workaround is to explicitly set `CFLAGS` to empty:
`./configure CFLAGS=`
=>
`CFLAGS:               -ffast-math -funroll-loops -fomit-frame-pointer -O3`

**Comment by umlaeute** [comment]
> it is really unfortunate if automake decides

yes, but automake fills these default values in each and every autotools project all over the world.

my take is that we do not want to divert from that standard.

**Comment by sebshader** [comment]
this is what I was trying to say here https://github.com/pure-data/pure-data/pull/1715#issuecomment-1207245294

**Comment by Spacechild1** [comment]
@umlaeute 

> yes, but automake fills these default values in each and every autotools project all over the world.

To be clear: I don't take issue with automake providing default values. But I would have expected that they'd be used for the project's *internal* CFLAGS and that a project would be able to override them on a per-configuration basis. Otherwise it would mean that a project cannot set its own optimization levels - which would be particularly absurd for debug configurations (you certainly don't want `-O2` for your debug build).

Now, https://autoconf.gnu.narkive.com/iOnZ6tU8/configure-sets-cflags-or-how-to-disable-default-cflags-g-o2-for-gcc seems to say that autotools really does fill missing `CFLAGS` with default options. I'm stumped.

On the other hand, we are not the only ones who want to set their own default optimization flags while letting the user override them with their `CFLAGS`. The linked thread above seems to contain some solutions, but I know too little about autotools to apply them...

**Comment by umlaeute** [comment]
> The linked thread above seems to contain some solutions

for me, the most obvious lines in this thread are:
> Very bad idea. 

> In short: You have opened a can of worms, you'd better avoid.

> Wrt. CFLAGS, there exist a lot of broken
configure scripts, which effectively outsmart themselves.

I'm aware, that all these remarks are from a single person...

**Comment by umlaeute** [comment]
so the question is, what we actually want.

i want:
- default to **working** build flags on any platform/compiler that is supported by autotools (with *working* defined as: creates a usable binary)
- adhere to standards (principle of least suprise)
- allow the users (who know what they are doing) to override the build flags to produce binaries that fit their personal requirements
- default to allowing for cross-compilation and out-of-tree builds.

others (e.g. @danomatika, @Spacechild1, @sebshader,...) might have different expectations. afaict, these are:
- simple way for the user to build a high-performance Pd (without forcing the user to read the docs for their compiler)
- simple way for the user to build a debuggable Pd (without forcing the user to read the docs for their compiler)

(all of these are probably gross simplifications.).

---

we currently try to add `DEBUG_CFLAGS` if configure was invoked with `--enable-debug`, and `RELEASE_CFLAGS` if not.

both `DEBUG_CFLAGS` and `RELEASE_CFLAGS`, make assumptions about the flags accepted by the compiler, but the latter (and default) does more so.
i also think they make assumptions about the performance gain, based solely on observations on the x86 and x86_64 platforms.

this implementation currently seems to not satisfy both parties:
- it fails to provide *working* build flags for all platforms provided by autotools (using build flags that are not known to work on compiler Y)
- it doesn't achieve the goal of simple build flags for debug or performance builds (as some of the flags are overwritten)
---

here's a suggestion: use three build-profiles, rather than two:
- debug builds
- normal builds
- performance builds

1. if debug-build is *explicitly* enabled with `--enable-debug`, set the CFLAGS to `$DEBUG_CFLAGS` (+ whatever the user provided when invoking configure), ignoring autotools default CFLAGS
2. if debug-build is *explicitly* disabled with `--disable-debug` (or `--enable-debug=no`), set the CFLAGS to `$RELEASE_CFLAGS` (+ whatever the user provided when invoking configure), ignoring autotools default CFLAGS
3. if debug-builds are neither explicitly enabled nor explicitly disabled, use the autotools defaults for CFLAGS (`-g -O2` or whatever)
4. regardless of which method was picked at configure-time, the can be completely overridden by passing it as a variable to the make-invocation  (`make CFLAGS="-O0"`)

this gets rid of the following problems:
- the default autoconf flags to longer override the hand-selected flags
- the default behavior is the well-known autotools behavior

the drawback is obviously, that you don't get a "performance build" by default, but you must manually enable it.

**Comment by Spacechild1** [comment]
I think we're conflating two issues here:

a) prevent autotools from adding *their* default compile options since we already provide our own

This is just a technical issue and I hope the linked thread contains some solutions.

b) offer more/better build configurations

I think your 3-build-profiles goes into the right direction, but I don't really like the semantics:

> if debug-builds are neither explicitly enabled nor explicitly disabled, use the autotools defaults for CFLAGS (-g -O2 or whatever)

How would a user go back to that configuration other than deleting the configure cache? I think we need something more explicit, e.g. `--enable-optimizations`.

Generally, I don't really like the idea of having "implicit" build types as a combination of configure options. What I really like about CMake is that it offers configurations per build type (https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html) which are filled with reasonable defaults, but can be overriden by the project *and* the user. I'm not sure if something like this can be applied to autotools in an idiomatic way. Probably not...

---

Anyway, I think this GitHub issue is really only about a) and we could use another issue to discuss b).

**Comment by umlaeute** [comment]
> a) prevent autotools from adding their default compile options since we already provide our own

my arguing is, that:
- our compile options are broken for any compilers other than gcc-compatible ones (gcc+clang)
- otoh, autotools only adds options it knows to work (e.g. it will only add the `-g -O2` defaults if it detects that the compiler can use them)

so i don't think we want to completely switch our flags for those suggested by autotools (the solution to that problem would actually be quite simple).

as for the need for "high performance builds": i understand that. and i don't think it is against the autotools philosophy in general.
what *is* against the philosophy is just randomly adding flags that happen to produce good results on the developers' system(s), without caring about other systems.

> How would a user go back to that configuration other than deleting the configure cache?

re-running `./configure`, but this time without the `--enable-debug` flag.
i'm not sure i see the problem. (yes, this is different from other build system where you can run `make Debug Release`; an autotool project would solve that with two separate out-of-tree builds)

---

as for `--enable-optimizations`: i quite like that, and was about to suggest that (but removed it again, in order to not add another new idea).
what would `--enable-debug` do then? (sidenote: shouldn't `--enable-debug` add something like `-DDEBUG` or `-D_DEBUG``)

**Comment by umlaeute** [comment]
> I think we need an explicit way to disable our optimization flags. This could be the above mentioned --disable-optimizations 

just to re-iterate: i do think it is a *very* bad idea to enforce these flags in the default builds.
so i suggest to do it the other way round: we need an explicit way to *enable* our optimization flags: `--enable-optimizations` would do nicely.

**Comment by Spacechild1** [comment]
> our compile options are broken for any compilers other than gcc-compatible ones (gcc+clang)

Which compilers are you thinking about? ICC? Then again, autotools does not even support one of the three major compilers (= MSVC)... If we want a truly cross-platform build system, we would have to move away from autotools anway.

> re-running ./configure, but this time without the --enable-debug flag.

Ah, my bad, I forgot how `configure` works... (I thought it caches the options between runs)

> just to re-iterate: i do think it is a very bad idea to enforce these flags in the default builds.
so i suggest to do it the other way round: we need an explicit way to enable our optimization flags: --enable-optimizations would do nicely.

I would be fine with that. As long as we document it prominently in the README.

> what would --enable-debug do then? (sidenote: shouldn't --enable-debug add something like -DDEBUG or `-D_DEBUG``)

It's the other way round: release builds shoudl add `-DNDEBUG`. I actually fixed that: https://github.com/pure-data/pure-data/commit/8c04895406eaa7ba0c98c61a33ac54f6adccd71c

**Comment by Spacechild1** [comment]
> what would --enable-debug do then?

I think `--enable-debug` would add `-g -O0` *after* `CFLAGS` as a quick way to get a proper debug build. `--disable-debug` would just add `-DNDEBUG`.

**Comment by umlaeute** [comment]
> It's the other way round: release builds should add `-DNDEBUG`. Actually, I already implemented that:

yes i know. but that is why i was asking: what is `--enable-debug` doing, as opposed to `--disable-debug`.

> EDIT2: but in that case, they could just do `CFLAGS="-g -Ox"` instead. Dunno.

well yeah. but in that case anybody who wanted optimized builds could just do `CFLAGS="-O10"`.
isn't this entire discussion about *not* doing this?

---

before talking about adding flags before or after, I would also like to recall, that there are actually two places where `CFLAGS` are set: at configure-time and at make-time.

we can do all kinds of things to the `CFLAGS` at configure-time (unsetting, setting, appending, prepending,...), but there are only two ways to inject these flags into the final makefile:
- either as direct `CFLAGS` which will always be the last cflags in the compiler-invocation, but which can always be overridden in its entirety by the user invoking `make CFLAGS="..."`
- or by `AM_CFLAGS` (or similar), which will *never* be the last cflags in the compiler-invocation.

```sh
$ ./configure CFLAGS="-DFOO"
[...]
$ make clean; make
gcc -DFOO -o foo foo.c
$ make clean; make CFLAGS="-DBAR"
gcc -DBAR -o foo foo.c
```

we are currently using the latter (`AM_CFLAGS`), which is why the default `CFLAGS` (from autoconf), override some of the flags we are setting.

it's probably simplest to use the former (ordinary `CFLAGS`), to which we can prepend at will.
the fact that the user can then completely undo these flags by running `make CFLAGS=""`.

i think this is reasonable behaviour, as those flags are not *required* flags anyhow...

**Comment by Spacechild1** [comment]
> what is --enable-debug doing

*not* passing `-DNDEBUG` :-)

**Comment by Spacechild1** [comment]
> well yeah. but in that case anybody who wanted optimized builds could just do CFLAGS="-O10".
isn't this entire discussion about not doing this?

Not really. Of course, power users can always pass their own flags. The discussion is really about providing reasonable defaults for creating release and debug builds because we can't really expect casual users to figure out the appropriate flags on their own.

**Comment by umlaeute** [comment]
>>  isn't this entire discussion about **not** doing this?
> Not really

[...]

>  The discussion is really about providing reasonable defaults.

yes. that's why i was questioning your suggestion of using passing `CFLAGS`.

did you miss the negation in my "**not**doing this"?

**Comment by Spacechild1** [comment]
See my post above. I was talking about a specific edge case for power users.

**Comment by Spacechild1** [comment]
> it's probably simplest to use the former (ordinary CFLAGS), to which we can prepend at will.
the fact that the user can then completely undo these flags by running make CFLAGS="".

Hmmm... but how would I, as a user, *add* options via `CFLAGS`?

**Comment by umlaeute** [comment]
> Hmmm... but how would I, as a user, add options via CFLAGS?

what would you like to add? why? and when?

if you are talking about `make CFLAGS=...`, then you are that deep into fuddling with the build system (aka: we are talking about "power users" right now), so what keeps you from manually adding your flags to the current list of `CFLAGS` in whatever order is fit and pass the entire shebang?

if you are talking about `./configure CFLAGS=...` then we *can* append/prepend/ignore/take these flags.

**Comment by Spacechild1** [comment]
@umlaeute Thanks for the clarification! I have been thinking about `./configure CFLAGS=...`, so there's no problem then.

**Comment by Spacechild1** [comment]
Side note:

It's quite interesting to see that autotools puts `CFLAGS` etc. at the end, while CMake puts them at the *beginning* (https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html).

That's how I came to understand it:

**Autotools**: users can, to some extend, override options by providing their own options with `./configure CFLAGS=...`. They explicitly acknowledge the limitations:
> It’s true that not all options to all programs can be overridden. So in general, users could conceivably want to place options at arbitrary places in the command line, but Automake does not support this. It would be difficult to make such generality comprehensible. Being able to specify the final options commonly suffices. 
https://www.gnu.org/software/automake/manual/html_node/Flag-Variables-Ordering.html

**CMake**: everything is a cache variable that the user can override. Consequently, if a user is not happy with the default configuration, they can set/change/remove the appropriate values directly.

**Comment by umlaeute** [comment]
so what do we want in the end?

i'll try to summarize:
- we are only modifying the `CFLAGS` as generated by `/configure` (and allow the user to completely override them when invoking `make`)

| `./configure --enable-...`  | `./configure CFLAGS=...` | used CFLAGS |
|------|--------|------|
|   |   | autotools default: `-g -O2`  |
|`--enable-debug` | | autotools default + special debug flags: `-g -O2 -g -O0` |
|`--enable-optimization` | |  autotools default + special release flags: `-g -O2 -O3` |
|`--enable-debug --enable-optimization` |  | autotools default + special debug flags + special release flags: `-g -O2 -g -O0 -O3` |
|  | `CFLAGS="-DFOO"` | user provided: `-DFOO` |
|`--enable-debug` | `CFLAGS="-DFOO"` | special debug flags + user provided: `-g -O0 -DFOO` |
|`--enable-optimization` | `CFLAGS="-DFOO"` | special release flags + user provided: `-O3 -DFOO` |
|`--enable-debug --enable-optimization` |`CFLAGS="-DFOO"` |  special debug flags + special release flags + user provided: `-g -O0 -O3 -DFOO` |

so the default autotools flags would always be prepended, and user-provided flags always be appended.

the reason why i suggest to always add the autotools flags (even if they seem redundant in the case of `--enable-debug` and adds `-g` even for `--enable-optimization`).

it seems the only reason to drop the `-g` flag is to keep the binary size small.
but a far more efficient way to do that is by just running `strip` on the final product once you are sure that you do not need any debugging symbols (at least on linux).

here's a size-comparision between two Pds that have been compiled once with `-g -O2` and once with just `-O2`:

```sh
$ ls -l
total 4972
-rwxr-xr-x 1 user user 3917352 Aug 30 15:59 pd.default
-rwxr-xr-x 1 user user 1167568 Aug 30 15:59 pd.nodebug
$ strip --strip-unneeded pd.*
$ ls -l
total 2016
-rwxr-xr-x 1 user user 1031472 Aug 30 15:59 pd.default
-rwxr-xr-x 1 user user 1031472 Aug 30 15:59 pd.nodebug
$ 
```

**Comment by Spacechild1** [comment]
Thanks for the summary! Looks good to me. (The release flags also contain `-ffast-math -funroll-loops -fomit-frame-pointer`, but you probably omitted it for brevity.)

> but a far more efficient way to do that is by just running strip on the final product once you are sure that you do not need any debugging symbols (at least on linux).

100% agree. I think autotools already provides a `install-strip` target that does exactly this. I wouldn't even mind shipping Pd with debug symbols.

> here's a size-comparision between two Pds that have been compiled once with -g -O2 and once with just -O2:

The size difference is a bit more pronounced on Windows because of the statically linked libraries. `pd.dll` is 6.1 MB unstripped and 1.6 MB stripped.

**Comment by danomatika** [comment]
An additional option is to define the default release and debug flags *per platform*.

My point of view is that the “normal build” without additional configure options should reflect that of Miller’s release build. By running make, users should a similar build with similar performance. I understand not wanting to fight the compiler but we want to make building Pd accessible without requiring too many caveats or options which may be obscure to many.

OTOH picking the available options such as sound backends is not always automatic, ala jack, so perhaps this is not realistic anyway?

In any case, if the default workflow is changed we just need to communicate effectively.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Comment by Spacechild1** [comment]
> My point of view is that the “normal build” without additional configure options should reflect that of Miller’s release build.

I agree with that and that's why I have suggested to use `--enable-optimization` by default. I think most users are just fine with our default optimization flags.

> OTOH picking the available options such as sound backends is not always automatic, ala jack, so perhaps this is not realistic anyway?

Also, on Windows you need to do `--enable-asio` (and download the ASIO SDK).

**Comment by Lucarda** [comment]
> On Windows you need to do --enable-asio (and download the ASIO SDK) and finally do make app. There is no way to just type ./configure; make and get a fully functional Pd :-)

@Spacechild1  are you sure?

```
./autogen.sh
./configure
make app
```

works for me (i mean I have ASIO) if I have placed the ASIO SDK before running the above commands.

**Comment by Spacechild1** [comment]
@Lucarda Ah, you're right, of course. ASIO support is enabled by default, but automatically disabled if the ASIO SDK is not present. But you still need to download the SDK :-)

**Comment by nioelumijkee** [comment]
disable adding flags by default, in the file configure.ac :

```
#########################################
##### OS Detection #####

...

#########################################
##### Default flags
${CFLAGS=""}
${CXXFLAGS=""}

#########################################
##### Check for programs, libs, & headers #####

...

```
this work with linux/gcc. (not tested on other platforms). flags set without default "-g -O2".

**Comment by umlaeute** [comment]
I'm still not convinced that this is a good idea in the first place

**Comment by nioelumijkee** [comment]
> I'm still not convinced that this is a good idea in the first place

in theory, this will simply break someone's system flags in -010.

**Comment by umlaeute** [comment]
no idea what `-010` means.


however, afaict your patch does two things:
- it completely overrides any CFLAGS (or CXXFLAGS) passed to the build system
- it breaks the build if you pass any CFLAGS (resp CXXFLAGS)

so: i don't see any improvement whatsoever, only two regressions.

**Comment by danomatika** [comment]
My preference is to enabled optimizations by default. If that is not wanted, then we should append this flag to the example build instructions in INSTALL.txt with an appropriate explanation that this is used by the Pd release builds. This is probably the appropriate section: https://github.com/pure-data/pure-data/blob/master/INSTALL.txt#L76
